Wikipedia/Code/WMFZeroConfigurationManager.m

```diff
@@ -0,0 +1,166 @@
+NSString *const WMFZeroXCarrier = @"X-Carrier";
+NSString *const WMFZeroXCarrierMeta = @"X-Carrier-Meta";
+
+@interface WMFZeroConfigurationManager ()
+
+@property (nonatomic, strong, readonly) WMFZeroConfigurationFetcher *zeroConfigurationFetcher;
+@property (nonatomic, strong, nullable, readwrite) WMFZeroConfiguration *zeroConfiguration;
+
+@property (atomic, copy, nullable) NSString* previousPartnerXCarrier;
+@property (atomic, copy, nullable) NSString* previousPartnerXCarrierMeta;
+@property (atomic, readwrite) BOOL isZeroRated;
+
+@end
+
+@implementation WMFZeroConfigurationManager
+@synthesize isZeroRated = _isZeroRated;
+@synthesize zeroConfigurationFetcher = _zeroConfigurationFetcher;
+
++ (void)load {
+    [super load];
+    //    FBTweakAction(@"Networking", @"Wikipedia Zero", @"Reset ZeroOnDialogShownOnce", ^{
+    //        [[NSUserDefaults wmf_userDefaults] setBool:NO forKey:WMFZeroOnDialogShownOnce];
+    //        [[NSUserDefaults wmf_userDefaults] synchronize];
+    //    });
+}
+
+- (WMFZeroConfigurationFetcher *)zeroConfigurationFetcher {
+    if (!_zeroConfigurationFetcher) {
+        _zeroConfigurationFetcher = [[WMFZeroConfigurationFetcher alloc] init];
+    }
+    return _zeroConfigurationFetcher;
+}
+
+- (void)setIsZeroRated:(BOOL)isZeroRated {
+    @synchronized(self) {
+        if(_isZeroRated != isZeroRated){
+            _isZeroRated = isZeroRated;
+            [[NSNotificationCenter defaultCenter] postNotificationName:WMFZeroRatingChanged object:self];
+        }
+    }
+}
+
+- (BOOL)isZeroRated {
+    BOOL isZeroRated;
+    @synchronized(self) {
+        isZeroRated = _isZeroRated;
+    }
+    return isZeroRated;
+}
+
+- (void)setWarnWhenLeaving:(BOOL)warnWhenLeaving {
+    [[NSUserDefaults wmf_userDefaults] setObject:[NSNumber numberWithBool:warnWhenLeaving]
+                                          forKey:WMFZeroWarnWhenLeaving];
+    [[NSUserDefaults wmf_userDefaults] synchronize];
+}
+
+- (BOOL)warnWhenLeaving {
+    return [[NSUserDefaults wmf_userDefaults] boolForKey:WMFZeroWarnWhenLeaving];
+}
+
+#pragma mark - Banner Updates
+
+/**
+ *   This method:
+ *
+ * - Fetches carrier specific strings placing them in self.zeroConfiguration object.
+ *
+ * - If the fetched zeroConfiguration has a nil "message" string, this means even though
+ *   there was a header, leading us to believe this network was Zero rated, its Zero
+ *   rating is not presently enabled. (It would be nice if the query fetching the
+ *   zeroConfiguration returned an "enabled" key/value, but it doesn't - it nils out the
+ *   values instead apparently.) So in the nil message case we set isZeroRated "NO".
+ */
+- (void)fetchZeroConfigurationAndSetIsZeroRatedIfNecessary {
+    
+    // Note: don't nil out self.zeroConfiguration in this method
+    // because if we do we can't show its exit message strings!
+
+    //TODO: ensure thread safety so we can do this work off the main thread...
```

Comment:
Not sure about this comment - what needs to move to a background thread? It doesn't look like any code on line 93 or 94 depends on the main thread.

Also line 99 is not guaranteed to be invoked on the main thread… which to me looks to be the only cause for concern of thread safety. 

In this simple case, you can always use `@synchronized(self){/*code that needs to be executed synchronously*/}` to make sure you do not access from multiple threads at the same time. I would recommend this approach for all the lines where you manipulate the `isZeroRated` ivar.

Related: Why do we need to re dispatch to the main queue below and then immediately invoking a promise network call (which immediately starts an asynchronous network call). I think this can be safely removed.